allennlp/modules/token_embedders/embedding.py

```diff
@@ -296,3 +352,183 @@ def _read_pretrained_hdf5_format_embedding_file(embeddings_filename: str, # pyli
+                    '.txt': io,
+                    '.vec': io,
+                    '.gz': gzip,
+                    '.bz2': bz2,
+                    '.lzma': lzma,
+                    }.get(extension, None)
+
+            if package is None:
+                logger.warning('The embeddings file has an unknown file extension "%s". '
+                               'We will assume the file is an (uncompressed) text file', extension)
+                package = io
+
+            self._handle = package.open(main_file_local_path, 'rt', encoding=encoding)  # type: ignore
+
+        # To use this with tqdm we'd like to know the number of tokens. It's possible that the
+        # first line of the embeddings file contains this: if it does, we want to start iteration
+        # from the 2nd line, otherwise we want to start from the 1st.
+        # Unfortunately, once we read the first line, we cannot move back the file iterator
+        # because the underlying file may be "not seekable"; we use itertools.chain instead.
+        first_line = next(self._handle)     # this moves the iterator forward
+        self._num_tokens = EmbeddingsTextFile._get_num_tokens_from_first_line(first_line)
+        if self._num_tokens:
+            # the first line is a header line: start iterating from the 2nd line
+            self._iterator = self._handle
+        else:
+            # the first line is not a header line: start iterating from the 1st line
+            self._iterator = itertools.chain([first_line], self._handle)
+
+    def _open_inside_zip(self, archive_path: str, member_path: Optional[str] = None) -> None:
+        cached_archive_path = cached_path(archive_path, cache_dir=self._cache_dir)
+        archive = zipfile.ZipFile(cached_archive_path, 'r')
+        if member_path is None:
+            members_list = archive.namelist()
+            member_path = self._get_the_only_file_in_the_archive(members_list, archive_path)
+        member_path = cast(str, member_path)
+        member_file = archive.open(member_path, 'r')
+        self._handle = io.TextIOWrapper(member_file, encoding=self._encoding)
+        self._archive_handle = archive
+
+    def _open_inside_tar(self, archive_path: str, member_path: Optional[str] = None) -> None:
+        cached_archive_path = cached_path(archive_path, cache_dir=self._cache_dir)
+        archive = tarfile.open(cached_archive_path, 'r')
+        if member_path is None:
+            members_list = archive.getnames()
+            member_path = self._get_the_only_file_in_the_archive(members_list, archive_path)
+        member_path = cast(str, member_path)
+        member = archive.getmember(member_path)   # raises exception if not present
+        member_file = cast(IO[bytes], archive.extractfile(member))
+        self._handle = io.TextIOWrapper(member_file, encoding=self._encoding)
+        self._archive_handle = archive
+
+    @property
+    def num_tokens(self) -> Optional[int]:
+        return self._num_tokens
```

Comment:
is there a good reason to have private properties with public getters? wouldn't it be simpler to just call the variables `num_tokens` and `uri` and let people access them directly? (that's how the rest of the codebase works)

Reply:
The reason is that in this way only the getter is public, not the setter. But I guess nobody would ever modify those fields. I can change that if you like it, for the sake of consistency.

Reply:
people could still modify `_num_tokens` if they wanted to that badly, I vote for just letting them have access to the field

Reply:
Sure, but people should know that modifying (or even accessing) a private field is a bad thing and if they do that is no more responsibility of the developer. Anyway, I'll change it :)